File: typescript/src/dialects/sql-dialect-base.ts
import type { SqlDialect } from "../sql.ts";
import type { ISqlDialect, SetupDdl } from "./sql-dialect.ts";

export abstract class SqlDialectBase implements ISqlDialect {
  abstract readonly name: SqlDialect;
  abstract readonly connectionEnvironmentVariables: readonly string[];
  readonly usesLastInsertRowid: boolean = false;
  readonly migratesDdl: string;
  readonly migrateLogsDdl: string;

  constructor(ddl: SetupDdl) {
    this.migratesDdl = ddl.migrates;
    this.migrateLogsDdl = ddl.migrateLogs;
  }

  quoteIdent(ident: string): string {
    return `"${ident}"`;
  }

  nowExpr(): string {
    return "CURRENT_TIMESTAMP";
  }

  limitClause(n: number): string {
    return `LIMIT ${n}`;
  }

  translatePlaceholders(sql: string): string {
    return sql;
  }

  async prerequisiteError(_connection: string): Promise<string | null> {
    return null;
  }

  async prepareSetup(_connection: string): Promise<void> {}

  abstract createClient(
    connection: string,
  ): Promise<import("../migration-client.ts").MigrationClient>;
}

File: csharp/Dialects/SqlDialectBase.cs
using System.Data.Common;

namespace Deterministic.MigrateRunner;

internal abstract class SqlDialectBase : ISqlDialect
{
    public abstract string Name { get; }
    public abstract IReadOnlyList<string> ConnectionEnvironmentVariables { get; }
    public abstract string MigratesDdl { get; }
    public abstract string MigrateLogsDdl { get; }
    public abstract bool UseTransaction { get; }

    public string SelectAppliedNamesSql =>
        $"SELECT {QuoteIdent("name")} FROM {QuoteIdent("migrates")}";

    public string SelectLastAppliedNameSql =>
        $"SELECT {QuoteIdent("name")} FROM {QuoteIdent("migrates")} ORDER BY {QuoteIdent("name")} DESC LIMIT 1";

    public string InsertAppliedSql =>
        $"INSERT INTO {QuoteIdent("migrates")} ({QuoteIdent("name")}, {QuoteIdent("checksum")}) VALUES ({Placeholder("p1")}, {Placeholder("p2")})";

    public string DeleteAppliedSql =>
        $"DELETE FROM {QuoteIdent("migrates")} WHERE {QuoteIdent("name")} = {Placeholder("p1")}";

    public abstract string NormalizeConnectionString(string connection);
    public abstract DbConnection CreateConnection(string connectionString);
    public abstract void AddParameter(DbCommand command, string name, object value);

    public virtual string? PrerequisiteError(string connection) => null;

    protected abstract string QuoteIdent(string ident);
    protected abstract string Placeholder(string name);

    protected static void Bind(DbCommand command, string parameterName, object value)
    {
        var p = command.CreateParameter();
        p.ParameterName = parameterName;
        p.Value = value;
        command.Parameters.Add(p);
    }
}
